core/trie/snap_support.go

```diff
@@ -0,0 +1,162 @@
+package trie
+
+import (
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+func (t *Trie) IterateAndGenerateProof(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error),
+) ([]ProofNode, bool, error) {
+	var lastKey *felt.Felt
+
+	finished, err := t.Iterate(startValue, func(key, value *felt.Felt) (bool, error) {
+		lastKey = key
+
+		return consumer(key, value)
+	})
+	if err != nil {
+		return nil, false, err
+	}
+
+	proofset := map[felt.Felt]ProofNode{}
+
+	// If start value is null && finished, you dont need to provide any proof at all
+	if !finished || startValue != nil {
+		feltBts := startValue.Bytes()
+		startKey := NewKey(t.height, feltBts[:])
+		// Yes, the left proof is actually for the start query, not the actual leaf. Very confusing, yea I know. Need to
+		// actually check that the server did not skip leafs.
+		leftProof, err := GetProof(&startKey, t)
+		if err != nil {
+			return nil, false, err
+		}
+		for _, proof := range leftProof {
+			// Well.. using the trie hash here is kinda slow... but I just need it to work right now.
+			proofset[*proof.Hash(t.hash)] = proof
+		}
+	}
+
+	if !finished && lastKey != nil {
+		feltBts := lastKey.Bytes()
+		lastKey := NewKey(t.height, feltBts[:])
+		rightProof, err := GetProof(&lastKey, t)
+		if err != nil {
+			return nil, false, err
+		}
+
+		for _, proof := range rightProof {
+			proofset[*proof.Hash(t.hash)] = proof
+		}
+	}
+
+	proofs := make([]ProofNode, 0, len(proofset))
+	for _, node := range proofset {
+		proofs = append(proofs, node)
+	}
+
+	return proofs, finished, nil
+}
+
+func VerifyRange(root, startKey *felt.Felt, keys, values []*felt.Felt, proofs []ProofNode, hash hashFunc,
+	treeHeight uint8,
+) (hasMore, valid bool, oerr error) {
+	proofMap := map[felt.Felt]ProofNode{}
+	for _, proof := range proofs {
+		proofHash := proof.Hash(hash)
+		proofMap[*proofHash] = proof
+	}
+
+	if len(proofMap) == 0 && startKey == nil {
+		// Special case where the whole trie is sent in one go.
+		// We just need to completely reconstruct the trie.
+
+		tempTrie, err := newTrie(newMemStorage(), treeHeight, hash)
+		if err != nil {
+			return false, false, err
+		}
+
+		for i, key := range keys {
+			_, err = tempTrie.Put(key, values[i])
+			if err != nil {
+				return false, false, err
+			}
+		}
+
+		recalculatedRoot, err := tempTrie.Root()
+		if err != nil {
+			return false, false, err
+		}
+
+		if !root.Equal(recalculatedRoot) {
+			return false, false, nil
+		}
+
+		return false, true, nil
+	}
+
+	if _, ok := proofMap[*root]; !ok {
+		// Verification failure, root not included in proof.
+		return false, false, nil
+	}
+
+	proofKeys := map[felt.Felt]Key{}
+	err := buildKeys(NewKey(0, []byte{}), root, proofMap, proofKeys, 0)
```

Comment:
This just builds the set of keys for all nodes from the root to the startKey, as outlined by the proofKeys path, right? If so I wonder if renaming proofKeys to ProofPathKeys might be a bit more clear?

Reply:
It might also be nice to define what the inputs are (eg keys, proofs, etc)

Reply:
Done